src/test/java/org/assertj/core/architecture/ArchitectureTest.java

```diff
@@ -0,0 +1,93 @@
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.junit.ArchUnitRunner;
+import com.tngtech.archunit.lang.ArchRule;
+import org.junit.runner.RunWith;
+
+import java.net.URL;
+import java.util.function.Predicate;
+
+import static com.tngtech.archunit.base.DescribedPredicate.alwaysTrue;
+import static com.tngtech.archunit.library.Architectures.layeredArchitecture;
+import static org.assertj.core.architecture.ArchitectureTest.ROOT_PACKAGE;
+
+@RunWith(ArchUnitRunner.class)
+@AnalyzeClasses(packages = ROOT_PACKAGE)
+public class ArchitectureTest {
+  static final String ROOT_PACKAGE = "org.assertj.core.";
+
+  private static final String ANNOTATIONS_LAYER = "annotations";
+  private static final String API_LAYER = "api";
+  private static final String CONDITION_LAYER = "condition";
+  private static final String CONFIGURATION_LAYER = "configuration";
+  private static final String DATA_LAYER = "data";
+  private static final String DESCRIPTION_LAYER = "description";
+  private static final String ERROR_LAYER = "error";
+  private static final String EXTRACTOR_LAYER = "extractor";
+  private static final String GROUPS_LAYER = "groups";
+  private static final String INTERNAL_LAYER = "internal";
+  private static final String MATCHER_LAYER = "matcher";
+  private static final String PRESENTATION_LAYER = "presentation";
+  private static final String UTIL_LAYER = "util";
+
+  private static final String API_PACKAGE = ROOT_PACKAGE + API_LAYER;
+
+  //@formatter:off
+  @ArchTest
+  private static final ArchRule layerRule = layeredArchitecture()
+    .layer(ANNOTATIONS_LAYER).definedBy(ROOT_PACKAGE + ANNOTATIONS_LAYER)
+    .layer(API_LAYER).definedBy(API_PACKAGE)
+    .layer(CONDITION_LAYER).definedBy(ROOT_PACKAGE + CONDITION_LAYER)
+    .layer(CONFIGURATION_LAYER).definedBy(ROOT_PACKAGE + CONFIGURATION_LAYER)
+    .layer(DATA_LAYER).definedBy(ROOT_PACKAGE + DATA_LAYER)
+    .layer(DESCRIPTION_LAYER).definedBy(ROOT_PACKAGE + DESCRIPTION_LAYER)
+    .layer(ERROR_LAYER).definedBy(ROOT_PACKAGE + ERROR_LAYER)
+    .layer(EXTRACTOR_LAYER).definedBy(ROOT_PACKAGE + EXTRACTOR_LAYER)
+    .layer(GROUPS_LAYER).definedBy(ROOT_PACKAGE + GROUPS_LAYER)
+    .layer(INTERNAL_LAYER).definedBy(ROOT_PACKAGE + INTERNAL_LAYER + ".(*)..")
+    .layer(MATCHER_LAYER).definedBy(ROOT_PACKAGE + MATCHER_LAYER)
+    .layer(PRESENTATION_LAYER).definedBy(ROOT_PACKAGE + PRESENTATION_LAYER)
+    .layer(UTIL_LAYER).definedBy(ROOT_PACKAGE + UTIL_LAYER)
+    .ignoreDependency(belongToTestClassDirectory(), alwaysTrue())
+
+    .whereLayer(INTERNAL_LAYER).mayOnlyBeAccessedByLayers(
+      ANNOTATIONS_LAYER,
+      API_LAYER,
+      CONDITION_LAYER,
+      CONFIGURATION_LAYER,
+      DATA_LAYER,
+      DESCRIPTION_LAYER,
+      ERROR_LAYER,
+      EXTRACTOR_LAYER,
+      GROUPS_LAYER,
+      MATCHER_LAYER,
+      PRESENTATION_LAYER,
+      UTIL_LAYER);
```

Comment:
I need to think about the dependencies rules of AssertJ (one is ideally `util` should not depend on `internal` but whether this is actually true or not is not sure at all).
I think a reasonable approach is to set the frame for architecture tests in this PR and add the rules AssertJ should follow in subsequent PR that would also refactor the code to enforce them.

Reply:
I modified the code to check the rule "util should not depend on internal" and the test pass.

Reply:
nice, thanks!

Reply:
Actually, due to bugs in the test, it didn't pass for the rule "util should not depend on internal" 
but now it's "fixed" by freezing the violation in archunit_store/a7253140-a775-4d8b-8f63-e57ad212fdea